src/main/test/java/io/burt/jmespath/contrib/function/JmesPathRuntimeWithStringFunctionTest.java

```diff
@@ -0,0 +1,586 @@
+   * https://www.w3.org/TR/xpath-functions-31/
+   * Those examples just illustrate the statements of specification rules and are considered as non-normative.
+   */
+  @Test
+  public void concatJoinsTheParts() {
+    functionRegistry.extend();
+    T result1 = search("concat('un', 'grateful')", dontCare);
+    T result2 = search("concat('Ingratitude, ', 'thou ', 'marble-hearted', ' fiend!')", dontCare);
+    assertThat(result1, is(jsonString("ungrateful")));
+    assertThat(result2, is(jsonString("Ingratitude, thou marble-hearted fiend!")));
+  }
+
+  @Test
+  public void concatFiltersOutNullTypes() {
+    T result1 = search("concat('Thy ', [], 'old ', `\"groans\"`, \"\", ' ring', ' yet', ' in', ' my', ' ancient',' ears.')", dontCare);
+    T result2 = search("concat('Ciao!',[])", dontCare);
+    assertThat(result1, is(jsonString("Thy old groans ring yet in my ancient ears.")));
+    assertThat(result2, is(jsonString("Ciao!")));
+  }
+
+  @Test
+  public void concatLiterals() {
+    T result1 = search("concat(`1`, `2`, `3`, `4`, `true`)", parse("{}"));
+    T result2 = search("concat(`1`, @)", parse("true"));
+    assertThat(result1, is(jsonString("1234true")));
+    assertThat(result2, is(jsonString("1true")));
+  }
+
+  @Test
+  public void concatRequiresAtLeastTwoArguments() {
+    thrown.expect(ParseException.class);
+    thrown.expectMessage(containsString("invalid arity calling \"concat\" (expected at least 2 but was 1)"));
+    search("concat(@)", parse("{}"));
+  }
+
+  @Test
+  public void lowerCaseExampleFromXPathSpec() {
+    T result = search("lower_case('ABc!D')", dontCare);
+    assertThat(result, is(jsonString("abc!d")));
+  }
+
+  @Test
+  public void lowerCaseRequiresASingleArgument() {
+    thrown.expect(ParseException.class);
+    thrown.expectMessage(containsString("invalid arity calling \"lower_case\" (expected 1 but was 2)"));
+    search("lower_case(@, @)", dontCare);
+  }
+
+  @Test
+  public void lowerCaseRequiresAStringAsArgument() {
+    thrown.expect(ArgumentTypeException.class);
+    thrown.expectMessage(containsString("expected string but was array"));
+    search("lower_case(@)", parse("[3]"));
+  }
+
+  @Test
+  public void upperCaseExamplesFromXPathSpec() {
+    T result = search("upper_case('abCd0')", dontCare);
+    assertThat(result, is(jsonString("ABCD0")));
+  }
+
+  @Test
+  public void upperCaseRequiresASingleArgument() {
+    thrown.expect( ParseException.class);
+    thrown.expectMessage(containsString("invalid arity calling \"upper_case\" (expected 1 but was 2)"));
+    search("upper_case(@, @)", dontCare);
+  }
+
+  @Test
+  public void upperCaseRequiresAStringAsArgument() {
+    thrown.expect(ArgumentTypeException.class);
+    thrown.expectMessage(containsString("expected string but was object"));
+    search("upper_case(@)", dontCare);
+  }
+
+  @Test
+  public void normalizeSpaceExamplesFromXPathSpec() {
```

Comment:
This tests looks superfluous, it tests all the aspects at once, and the subsequent tests test each aspect in isolation. That means that every time one of the other tests fail, this will fail too, and if this test fails at least one other test will fail – in other words, it adds no information. 

I'd rather see the failure "normalizeSpaceRemovesLeadingWhitespaces" than the failures "normalizeSpaceExamplesFromXPathSpec" and "normalizeSpaceRemovesLeadingWhitespaces". The former list (with the one) tells me what's wrong, the latter (with two) makes me confused as to what else but leading whitespace is wrong.

Reply:
Totally agreed. I remove it.

Back that time it seemed me important to add each and every non-normative example from XPath specification. Later on I split up most of them to provide better insight what should really happen.